「Guide > Scaling Up > Server-Side Rendering (SSR)」の翻訳 (#520)
vuejs-translations/docs-ja: src/guide/scaling-up/ssr.md

outline: deep
---

# サーバーサイドレンダリング (SSR)

## 概要

### SSR とは？

> Vue.js は、クライアントサイドのアプリケーションを構築するためのフレームワークです。デフォルトでは、Vue コンポーネントは、ブラウザー内の出力としてコンポーネントを生成し、DOM を操作します。ですが、同じコンポーネントをサーバー上で HTML 文字列にレンダリングし、それをブラウザーに直接送信し、最終的にクライアント上で完全にインタラクティブなアプリケーション内に静的なマークアップを"hydrate（ハイドレート）"することもできます。

Author: ありがとうございます！
提案もらった文章の方が分かりやすいですね、修正します。

サーバーレンダリングの Vue.js アプリケーションは、アプリケーションのコードの大部分がサーバー**と**クライアントの両方で実行されるという意味で、"isomorphic"（アイソモルフィック）または "universal"（ユニバーサル）であると考えることもできます。

### なぜ SSR なのか？

クライアントサイドの SPA（シングルページアプリケーション）と比較して、主な SSR の利点は次の点にあります:

> - **コンテンツ表示速度の高速化**: これは、低速のインターネットや低速のデバイスでより顕著です。サーバーレンダリングのマークアップは、すべての JavaScript のダウンロードと実行が完了するまで表示されるのを待つ必要がないため、ユーザーはより早く、完全にレンダリングされたページを見ることができます。さらに、初回アクセス時のデータ取得はサーバーサイドで行われるため、クライアントよりもデータベースへの接続が高速になる可能性があります。これにより、一般的には [Core Web Vitals](https://web.dev/vitals/) メトリクスが向上し、ユーザー体験が改善されます。また、コンテンツ表示速度がコンバージョン率に直結するアプリケーションでは重要です。

Author: `time-to-content　`
`time`なので、直訳ならコンテンツ表示時間となりますが、日本では広くコンテンツ表示速度と呼ばれているため、この訳にしていました。

- **統一されたメンタルモデル**: バックエンドのテンプレートシステムとフロントエンドのフレームワークを行ったり来たりする代わりに、同じ言語で、同じ宣言型の、コンポーネント指向のメンタルモデルを使って、アプリケーション全体を開発することができます。

- **SEO 向上**: 検索エンジンクローラーが完全にレンダーされたページを直接見ることができます。

  :::tip

>   現時点で Google と Bing は、同期型の JavaScript アプリケーションを問題なくインデックスできます。ここでのキーワードは同期です。アプリケーションがローディングスピナー（ロード状態）と共に始まり、Ajax でコンテンツを取得する場合、クローラーはアプリケーションの終了を待ちません。つまり、SEO が重要なページで非同期にコンテンツを取得する場合、SSR が必要になるかもしれません。

Author: `loading spinner`
直訳だと意味がわかりにくいかと思い、補足を入れています

  :::

また、SSR を使用する際には、いくつかのトレードオフを考慮する必要もあります:

- 開発の制約。ブラウザー固有のコードは、特定のライフサイクルフック内でのみ使用可能です。また、外部ライブラリーの中には、サーバーレンダリング時のアプリケーション内で実行できるようにするためには特別な処理が必要なものがあります。

- ビルドのセットアップとデプロイの要件がより複雑になります。任意の静的ファイルサーバーにデプロイできる完全静的 SPA とは異なり、サーバーレンダリングアプリケーションは Node.js サーバーを実行できる環境が必要です。

- サーバー側の負荷が増えます。Node.js で完全なアプリケーションをレンダリングすると、静的なファイルだけを提供するよりも CPU 負荷が高くなります。したがって、高トラフィックが予想される場合は、相応のサーバー負荷に備え、賢いキャッシュ戦略をとる必要があります。

> アプリケーションに SSR を用いる前に、まず問うべき質問は、それが本当に必要かどうかということです。これは、アプリケーションにとってコンテンツ表示速度がどれだけ重要であるか次第です。例えば、チーム内のダッシュボードを作成していて、数百ミリ秒の初期ロードがあってもそれほど重要ではない場合、SSR はオーバーエンジニアリングになります。ですが、コンテンツ表示速度が絶対的に重要であるケースにおいては、SSR は最大限の初期ロードパフォーマンスの達成に役立ちます。

Reviewer: よいと思います！ 👍

> アプリケーションに SSR を用いる前に、まず問うべき質問は、それが本当に必要かどうかということです。これは、アプリケーションにとってコンテンツ表示速度がどれだけ重要であるか次第です。例えば、チーム内のダッシュボードを作成していて、数百ミリ秒の初期ロードがあってもそれほど重要ではない場合、SSR はオーバーエンジニアリングになります。ですが、コンテンツ表示速度が絶対的に重要であるケースにおいては、SSR は最大限の初期ロードパフォーマンスの達成に役立ちます。

Author: 確かにそうですね、修正します。

### SSR vs. SSG

> **スタティックサイトジェネレーション (SSG)** は、プリレンダリングとも呼ばれ、高速なウェブサイトを構築する、もう 1 つのポピュラーな技術です。ページのサーバーレンダリングに必要なデータが、全てのユーザーに対して同じであれば、リクエストが来るたびにページをレンダリングする代わりに、あらかじめビルド時に一度だけレンダリングしておくことができます。プリレンダリングされたページは、静的な HTML ファイルとして生成・提供されます。

Reviewer: `Static Site` はここ最近 web 関連の front-end 系のブログ等で `静的サイト` と書いてあることがあるので、そのように訳してもよさそうです。

Reviewer: `popular technique` は良く使われていることを意味しているので、カタカナ日本語ではなく`代表的な技術` と意訳気味にしてもよい気がします。

Author: こちら確認もれてましてすみません、
修正します！

> SSG は SSR アプリケーションと同じパフォーマンス特性を持っており、優れたコンテンツ表示速度性能を提供します。同時に、出力内容が静的な　HTML　とアセットであるため、SSR　アプリケーションよりも低コストで簡単なデプロイが可能です。ここで重要なのは **static (静的)** というキーワードです: SSG は静的なデータ、つまりビルド時に既に内容が分かっていて、デプロイ時に変更されないものデータを使用するページにのみ適用できます。よって、データが変更されるたびに、新しいデプロイが必要です。

Author: 失礼しました、修正します。

> もしあなたが一握りのマーケティングページ（例：`/` 、`/about` 、`/contact` など）の SEO を改善するためだけに SSR を調べているなら、おそらく SSR の代わりに SSG が適しているでしょう。SSG は、ドキュメントサイトやブログのようなコンテンツベースのウェブサイトにも適しています。実際、あなたが今読んでいるこのウェブサイトは、Vue を利用した静的サイトジェネレーター [VitePress](https://vitepress.vuejs.org/) を使って静的に生成されています。

Author: `you probably want SSG instead of SSR`
`望む`、`必要となる`、と訳すと不自然なため意訳してにしていました。

Reviewer: そのまま訳すと確かにそうなので、意訳気味でいいかと思います！:+1:

> ## ベーシックチュートリアル

Reviewer: `Basic` は以下のように `基本` と訳してもよさそうです。
```suggestion
## 基本チュートリアル
```

Author: 確かにそうですね、修正します！

})
```

> 実行するには以下のようにします:

Reviewer 2: ```suggestion
次に以下を実行します:
```

Author: 修正します！

}
```

このファイルと依存関係は、サーバーとクライアントの間で共有されます。これを **universal code（ユニバーサルコード）** と呼びます。ユニバーサルコードを書くときには、いくつか注意しなければならないことがありますが、それについては[後述します](#writing-ssr-friendly-code)。

> クライアントエントリーは、ユニバーサルコードを取り込み、アプケーションを作成し、マウントを実行します:

Reviewer 2: ```suggestion
クライアントエントリーは、ユニバーサルコードをインポートしてアプリケーションを作成し、マウントを実行します:
```

Author: こちらも修正します！

})
```

さらに、ブラウザーでクライアントのファイルを読み込むためには、次のようなことも必要です:

1. `server.js` に `server.use(express.static('.'))` を追加して、クライアントにファイルを提供する。
2. HTML シェルに `<script type="module" src="/client.js"></script>` を追加して、クライアントエントリーを読み込む。
3. HTML シェルに [Import Map](https://github.com/WICG/import-maps) を追加して、ブラウザー内で `import * from 'vue'` のような用法をサポートする。

[StackBlitz で完璧な使い方を試す](https://stackblitz.com/fork/vue-ssr-example?file=index.js)。いま、こちらのボタンはインタラクティブになっています！

> ## 上位レベルの解決法

Reviewer 2: 「上位」だと若干ニュアンスが変わってしまうので変更
```suggestion
## より高度な解決法
```

Author: 上位だと優劣を感じさせてしまいますね😅修正します。

この例から本番用の SSR アプリケーションに移行するには、さらに多くのことが必要になります。必要なこととしては:

- Vue の単一ファイルコンポーネントやその他のビルドステップが必要とするもののサポート。実際は、同じアプリケーションに対して 2 つのビルドを調整する必要があります: 1 つはクライアント、もう 1 つはサーバー側です。

  :::tip
  SSR に使用される場合、Vue コンポーネントは、異なる方法でコンパイルされます。より効率的なレンダリングパフォーマンスを発揮するために、テンプレートは仮想 DOM レンダー関数の代わりに、連結した文字列にコンパイルされます。
  :::

- サーバー内のリクエストハンドラーでは、正しいクライアント側のアセット・リンクと最適なリソース・ヒントを共に HTML をレンダリングしてください。また、SSR モードと SSG モードを切り替えたり、同じアプリケーション内で両方を混在させる必要がある場合もあります。

- ルーティング、データ取得、状態管理ストアはユニバーサルな方法で管理します。

> 完全な実装をするのは非常に複雑で、かつ、あなたが選んだビルドツールチェーンに依存します。したがって、私たちの考えた、複雑性を抽象化し、上位レベルのソリューションを利用することを強くお勧めします。以下では、Vue のエコシステムで推奨されるいくつかの SSR ソリューションを紹介します。

Author: `opinionated solution`
調べてみたところ`自説を固執する`、`お気に入りの`などの訳があったのですが、しっくりくる訳がなかったので、
`vueコミュニティが推奨する解決策`というニュアンスを含んだ意訳にしました。

Reviewer: いいと思います！ 👍

SSR では、各リクエストの URL はアプリケーションにとって望ましい状態にマッピングされます。ユーザーとのインタラクションも DOM の更新もないので、サーバー側でリアクティビティは不要です。デフォルトでは、パフォーマンス向上のために SSR 中のリアクティビティは無効化されています。

### コンポーネントのライフサイクルフック

動的な更新がないため、<span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span> や <span class="options-api">`updated`</span><span class="composition-api">`onUpdated`</span> などのライフサイクルフックは SSR の間に**呼ばれず**、クライアント側でのみ実行されます。<span class="options-api"> SSR 中に呼ばれるフックは `beforeCreate` と `created` だけです。</span>

<span class="options-api">`beforeCreate` や `created` </span><span class="composition-api">`setup()` あるいは `<script setup>` のルートスコープ</span>でクリーンアップが必要になるような副作用を発生させるようなコードは避けるべきでしょう。このような副作用の例としては、`setInterval` を使用してタイマーを設定することが挙げられます。クライアントサイドのみのコードでは、タイマーをセットアップして、<span class="options-api">`beforeUnmount`</span><span class="composition-api">`onBeforeUnmount`</span> もしくは <span class="options-api">`unmounted`</span><span class="composition-api">`onUnmounted`</span>で終了させることがあります。ですが、SSR の間にアンマウントフックが呼ばれることはないため、タイマーはずっと残ることになります。代わりに、これを避けるために、副作用のあるコードを <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span> 内に移動しましょう。

### 特定のプラットフォームの API にアクセスする

> ユニバーサルコードはプラットフォーム固有の API へのアクセスを想定できないため、`window` や `document` のようなブラウザー専用のグローバルクラスを直接使用したコードは、Node.js で実行するとエラーが発生します。また、その逆も同様です。

Author: `global`
グローバルとだけ訳すと分かりにくいため`クラス`を付けています

Reviewer: `クラス` はこの文脈では使わずに、そのまま `グローバル` でよい気がします。
`クラス` は web では、 DOM の `class` や、JavaScript / プログラミング言語の `class` といったり、様々な方面で使われているので。

Author: 指摘いただいた通りですね。。。修正します！

プラットフォームの API が異なる状態で、サーバーとクライアントで共有するタスクがある場合は、ユニバーサル API 内でプラットフォーム固有の API を実装してラップするか、それを行ってくれるライブラリーを使用することを推奨します。例えば、[`node-fetch`](https://github.com/node-fetch/node-fetch) を使用すると、サーバーとクライアントの両方で同じフェッチ API を使用することができます。

ブラウザー専用 API の場合、一般的なアプローチは、<span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span>のようなクライアントだけのライフサイクルフックの中で遅延アクセス（lazy access）することです。

> サードパーティのライブラリーがユニバーサルでの使用を念頭に置いて書かれていない場合、サーバーレンダリングされたアプリケーションに統合するのは困難な可能性があることに注意してください。グローバルクラスのいくつかをモックすることで動作させることができるかもしれません。しかし、それは場当たり的な対応に過ぎず、他のライブラリーの環境検出コードに干渉してしまう可能性があります。

Author: `hacky`
スラングでいろいろな訳し方があるようでしたが、[こちら](https://hinative.com/ja/questions/13824464)を参考に訳にしてみました。

Reviewer: `hacky` の訳はいいですね！ 👍 

ここでも、`グローバルクラス` という訳がありますが、`グローバル` でよさそうです。

Author: ありがとうございます！
こちらも指摘いただいた通りなので修正します！

### クロスリクエスト状態汚染

状態管理のチャプターでは、[Reactivity API を使ったシンプルな状態管理パターン](state-management.html#simple-state-management-with-reactivity-api)を紹介しました。SSR においては、このパターンにいくつかの追加的な調整が必要です。

このパターンでは、JavaScript モジュールのルートスコープで共有する状態を宣言します。これはそれらの状態を**シングルトン**、つまり、アプリケーションのライフサイクル全体を通じて 1 つだけのリアクティブオブジェクトのインスタンスにします。これは純粋なクライアントサイドの Vue アプリケーションで期待通りに動作します。なぜなら、アプリケーションのモジュールは、ブラウザーがページへアクセスする度に初期化されるからです。

しかし、SSR のコンテキストでは、アプリケーションモジュールは通常、サーバーが起動したときに一度だけ初期化されます。同じモジュール・インスタンスが複数のサーバー・リクエストにまたがって再利用されるので、シングルトンの状態オブジェクトも同じように再利用されます。あるユーザーに固有のデータを含む共有シングルトンの状態を変更した場合、誤って情報が他のユーザーからのリクエストに漏れてしまうことがあります。これを **cross-request state pollution （クロスリクエスト状態汚染）** と呼びます。

技術的には、ブラウザーで行うように、リクエストごとにすべての JavaScript モジュールを再初期化することも可能です。しかし、JavaScript モジュールの初期化にはコストがかかるため、サーバーのパフォーマンスに大きな影響を与えることになります。

> 推奨される解決策は、ルーターやグローバルストアを含むアプリケーション全体の新しいインスタンスを、リクエストごとに作成することです。それから、コンポーネントで直接インポートするのではなく、[アプリケーションレベルの Provide](/guide/components/provide-inject.html#app-level-provide) を使って共有する状態を提供し、それを必要とするコンポーネントでインジェクトすることです:

Author: リンクテキストはリンク先となる[Provide/Inject](https://github.com/vuejs-translations/docs-ja/pull/466)の翻訳と合わせています。

```js
// app.js （サーバーとクライアントで共有されます）
import { createSSRApp } from 'vue'
import { createStore } from './store.js'

export function createApp() {
  const app = createSSRApp(/* ... */)
  // リクエスト単位で状態の新しいインスタンスを作成します
  const store = createStore(/* ... */)
  // 状態をアプリケーションレベルで提供します
  app.provide('store', store)
  // ハイドレーション目的で状態を公開することもできます
  return { app, store }
}
```

Pinia のような状態管理ライブラリーは、このことを念頭に置いて設計されています。詳しくは[Pinia の SSR ガイド](https://pinia.vuejs.org/ssr/)を参照してください。

### ハイドレーション・ミスマッチ

> プリレンダリングされた HTML の DOM 構造がクライアント側のアプリケーションが期待する内容と異なったときは、ハイドレーション・ミスマッチエラーになります。多くのケースでは、ブラウザーがネィティブ HTML を無効な HTML 文字列を正しい構造にパースしようとした時の動作によって起こります。 例えば、一般的な例として [`<div>` は `<p>` の内部に配置できない](https://stackoverflow.com/questions/8397852/why-cant-the-p-tag-contain-a-div-tag-inside-it)ことなどがあります:

Reviewer 2: ```suggestion
プリレンダリングされた HTML の DOM 構造がクライアント側のアプリケーションが期待する内容と異なったときは、ハイドレーション・ミスマッチエラーになります。多くのケースでは、ブラウザーのネイティブな HTML パース処理が、HTML 文字列の無効な構造を修正しようとした時に起こります。 例えば、一般的な例として [`<div>` は `<p>` の内部に配置できない](https://stackoverflow.com/questions/8397852/why-cant-the-p-tag-contain-a-div-tag-inside-it)ことなどがあります:
```

Author: こちらも修正します！
